### PR TITLE
[RW-4258][RISK:NO] Admin UI for CRUD of Institute entities for affiliation - Add/Edit

### PR DESCRIPTION
**Pull request by NehaBroad**
Screen Shot 2020-06-02 at 9 40 05 AM

IF back button is selected after editing but not saving
Screen Shot 2020-06-02 at 9 41 31 AM

Screen Shot 2020-06-02 at 9 40 37 AM
Screen Shot 2020-06-05 at 4 10 46 PM

**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI + API server with `yarn test-local`
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java [hunk 1]
```diff
@@ -128,20 +138,25 @@ public void deleteInstitution(final String shortName) {
...
+                // an update
+
+                final DbInstitution dbObjectToUpdate =
+                    institutionDao.save(
+                        institutionMapper.modelToDb(institutionToUpdate).setInstitutionId(dbId));
+                populateAuxTables(institutionToUpdate, dbObjectToUpdate);
+                return toModel(dbObjectToUpdate);
+              });
+    } catch (DataIntegrityViolationException ex) {
+      throw new ConflictException("Display Name already exist");
```
Comment (NehaBroad) [inline api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java:158]: The assumption is that request will be coming from UI ,the shortName will be different as we will be generating it  and just display Name is something we will not have control over and can be same.

File: api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java [hunk 2]
```diff
@@ -260,17 +275,36 @@ public void test_nonUniqueEmailPatterns() {
         .containsExactlyElementsIn(similarInstViaDb.getEmailAddresses());
   }
 
-  @Test
+  // Display Name cannot be empty
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java:278]: That's not what this test is intended to show.  Please update the `new Institution()` to supply a valid Institution.

File: api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java [hunk 3]
```diff
@@ -260,17 +275,36 @@ public void test_nonUniqueEmailPatterns() {
...
+            .shortName("test")
+            .displayName("We are all individuals")
+            .organizationTypeEnum(OrganizationType.INDUSTRY));
+    service
+        .createInstitution(new Institution().shortName("test").displayName("I'm not"))
+        .organizationTypeEnum(OrganizationType.EDUCATIONAL_INSTITUTION);
+  }
+
+  @Test(expected = ConflictException.class)
+  public void test_nonUniqueDisplayName() {
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java:298]: 👍

File: api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java [hunk 4]
```diff
@@ -511,6 +558,64 @@ public void validate_OperationalUser_nullInstitution() {
     assertThat(service.validateOperationalUser(institution)).isFalse();
   }
 
+  @Test(expected = BadRequestException.class)
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java:561]: I like these.  Could you add one more for enforcing that OtherText exists when the type is OTHER?

File: api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java (hunk 2 above)
Comment (NehaBroad) [reply api/src/test/java/org/pmiops/workbench/institution/InstitutionServiceTest.java:278]: Ahh i see its trying to verify updateInstitution with valid institute but a shortName thats not in DB

File: api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java [hunk 5]
```diff
@@ -269,6 +284,13 @@ private void populateAuxTables(
           new InstitutionUserInstructions()
               .institutionShortName(modelInstitution.getShortName())
               .instructions(userInstructions));
+    } else {
+      // Remove institution entry from institution_user_instructions table if user_instructions is
```
[the remark this answers is not in the document]
Comment (NehaBroad) [reply api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java:288]: I have already added test for this scenario test_updateInstitution_RemoveUserInstructionFromExistingInstitution

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+interface InstitutionEditState {
+  apiErrorMsg: string;
+  isAddInstitution: boolean;
+  institution: Institution;
+  invalidEmailAddress: boolean;
+  invalidEmailAddressMsg: string;
+  invalidEmailDomain: boolean;
+  invalidEmailDomainsMsg: string;
+  showOtherInstitution: boolean;
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:42]: This makes me think it means "show other institutions" so we should rename it to something which makes it clear it means "show the other-text field"

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    } else {
+      title = 'Add new Institution';
+      this.setState({isAddInstitution: true});
+    }
+
+  }
+
+  get institutionTypeOptions() {
+    const options = [
+      {label: 'Industry', value: OrganizationType.INDUSTRY},
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:90]: these should match the labels in admin-institution - let's make them use a common source.

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+      {label: 'Academic Research Institution', value: OrganizationType.ACADEMICRESEARCHINSTITUTION},
+      {label: 'Educational Institution', value: OrganizationType.EDUCATIONALINSTITUTION},
+      {label: 'Health Center non profit', value: OrganizationType.HEALTHCENTERNONPROFIT},
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:93]: Update this to "Health Center / Non-Profit"

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+    ];
+    return options;
+  }
+
+  // Filter out empty line or empty email addresses like <email1>,,<email2>
+  // Confirm each email is a valid email using validate.js
+  validateEmailAddresses() {
+    const invalidEmailAddress = [];
+    const {emailAddresses} = this.state.institution;
+    this.state.institution.emailAddresses = emailAddresses.filter(
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:112]: is this a safe way to set state?  We should probably use a local variable here and call setState() to update this list when we're done - or maybe it would be better not to update the state of `emailAddresses` at all?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+    if (invalidEmailAddress.length > 0) {
+      const errMessage = 'Following Email Addresses are not valid : ' + invalidEmailAddress.join(' , ');
+      this.setState({invalidEmailAddressMsg: errMessage});
+    }
+    return;
+  }
+
+  // Filter out empty line or empty email addresses like <email1>,,<email2>
+  // Confirm each email domain matches with regex
+  validateEmailDomain() {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:137]: plural: domains

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    });
+    this.setState({invalidEmailAddress: invalidEmailAddress.length > 0});
+    if (invalidEmailAddress.length > 0) {
+      const errMessage = 'Following Email Addresses are not valid : ' + invalidEmailAddress.join(' , ');
+      this.setState({invalidEmailAddressMsg: errMessage});
+    }
+    return;
+  }
+
+  // Filter out empty line or empty email addresses like <email1>,,<email2>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:135]: remove this line

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    });
+    this.setState({invalidEmailDomain: invalidEmailDomain.length > 0});
+    if (invalidEmailDomain.length > 0) {
+      const errMessage = 'Following Email Domains are not valid : ' + invalidEmailDomain.join(' , ');
+      this.setState({invalidEmailDomainsMsg: errMessage});
+    }
+    return;
+  }
+
+  setEmailDomain(emailDomains, attribute) {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:159]: plural

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+
+  // Check if the fields have not been edited
+  fieldsNotEdited() {
+    return (this.state.isAddInstitution && !this.fieldsNotEditedAddInstitution)
+        || (institutionToEdit && this.fieldsNotEditedEditInstitution);
+  }
+
+  get fieldsNotEditedAddInstitution() {
+    const {institution} = this.state;
+    return institution.displayName || institution.organizationTypeOtherText ||
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:172]: missing userInstructions

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+        institution.organizationTypeEnum === institutionToEdit.organizationTypeEnum &&
+        institution.duaTypeEnum === institutionToEdit.duaTypeEnum &&
+        institution.emailAddresses === institutionToEdit.emailAddresses &&
+        institution.emailDomains === institutionToEdit.emailDomains &&
+        institution.userInstructions === institutionToEdit.userInstructions &&
+        institution.organizationTypeOtherText === institutionToEdit.organizationTypeOtherText;
+  }
+
+
+  noEmptyRequiredFields() {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:188]: how about validateRequiredFields?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+    const {institution} = this.state;
+    let emailValid = true;
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:190]: this shouldn't ever be relevant, but false is safer here

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+        (institution.organizationTypeEnum === OrganizationType.OTHER &&
+            !institution.organizationTypeOtherText);
+  }
+
+  // Disable save button if
+  // a) No fields were edited or if there are any errors
+  // b) email address/Domain are not valid
+  // c) Required fields are not empty
+  disableSave(errors) {
+    return this.noEmptyRequiredFields() || (errors && errors.displayName) || this.fieldsNotEdited()
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:206]: why do we need to check displayName specifically?  Isn't "errors" enough?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+      || this.state.invalidEmailAddress || this.state.invalidEmailDomain;
+  }
+
+  async saveInstitution() {
+    const {institution} = this.state;
+    if (institution) {
+      this.setState({invalidEmailAddress: false});
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:213]: why is this necessary?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+      await institutionApi().createInstitution(institution)
+        .then(value => this.backNavigate())
+        .catch(reason => this.handleError(reason));
+    }
+  }
+
+  handleError(rejectReason) {
+    let errorMsg = 'Error while saving Institution. Please try again later';
+    if (rejectReason.status === 409) {
+      errorMsg  = 'Institution with Name ' + this.state.institution.displayName + ' already exist';
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:237]: We can't say for certain that this is the reason.  Better to use the error text which is returned here.

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 19]
```diff
@@ -0,0 +1,411 @@
...
+    } else {
+      this.backNavigate();
+    }
+  }
+
+  backNavigate() {
+    navigate(['admin/institution']);
+  }
+
+  isEmailAddressValid() {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:258]: are these two checking for validity or presence?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+            </TooltipTrigger>
+          </div>
+        </FlexRow>
+        <FlexRow>
+          <FlexColumn style={{width: '50%'}}>
+            <TextInputWithLabel
+                value={fp.startCase(institution.displayName)}
+                inputId='displayName'
+                inputName='displayName'
+                placeholder='New Username'
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:322]: fix

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+                inputStyle={{width: '16rem', marginTop: '0.3rem'}}
+                labelText='Institution Name'
+                onChange={v => this.setState(fp.set(['institution', 'displayName'], v))}
+            />
+            <div style={{color: colors.danger}} data-test-id='displayNameError'>
+              {!isAddInstitution && errors && errors.displayName}
+              </div>
+            <label style={styles.label}>Institution Type</label>
+            <Dropdown style={{width: '16rem'}} data-test-id='role-dropdown'
+                      placeholder='Your Role'
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:333]: fix

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 22]
```diff
@@ -0,0 +1,411 @@
...
+               onChange={v => this.setState(fp.set(['institution', 'organizationTypeOtherText'], v))}
+               inputStyle={{width: '16rem', marginTop: '0.8rem'}}/>}
+            <label style={styles.label}>Agreement Type</label>
+            <Dropdown style={{width: '16rem'}} data-test-id='agreement-dropdown'
+                      placeholder='Your Agreement'
+                      options={this.institutionAgreementTypeOptions}
+                      value={institution.duaTypeEnum}
+                      onChange={v => this.setState(fp.set(['institution', 'duaTypeEnum'], v.value))}/>
+            {institution.duaTypeEnum === DuaType.RESTRICTED && <FlexColumn data-test-id='emailAddress' style={{width: '16rem'}}>
+              <label style={styles.label}>Accepted Email Address</label>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:349]: plural:Addresses

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 23]
```diff
@@ -0,0 +1,411 @@
...
+              <TextArea value={institution.emailAddresses && institution.emailAddresses.join(',\n')}
+                        data-test-id='emailAddressInput'
+                        onBlur={(v) => this.validateEmailAddresses()}
+                  onChange={(v) => this.setEmailDomain(v, 'emailAddresses')}/>
+              {this.state.invalidEmailAddress && <div data-test-id='emailAddressError' style={{color: colors.danger}}>
+                {this.state.invalidEmailAddressMsg}
+                </div>}
+            </FlexColumn>}
+            {institution.duaTypeEnum === DuaType.MASTER && <FlexColumn data-test-id='emailDomain' style={{width: '16rem'}}>
+              <label style={styles.label}>Accepted Email Domain</label>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:359]: plural: Domains

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+              <TextArea value={institution.emailDomains && institution.emailDomains.join(',\n')} onBlur={(v) => this.validateEmailDomain()}
+                        data-test-id='emailDomainInput'
+                        onChange={(v) => this.setEmailDomain(v, 'emailDomains')}/>
+              {this.state.invalidEmailDomain && <div data-test-id='emailDomainError' style={{color: colors.danger}}>
+                {this.state.invalidEmailDomainsMsg}
+                </div>}
+            </FlexColumn>}
+          </FlexColumn>
+          <FlexColumn style={{width: '50%', marginRight: '1rem'}}>
+            <label style={{...styles.label, marginTop: '0rem'}}>User instruction (Optional)</label>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:369]: Plural: instructions

Maybe something more explicit would be appropriate, like "User Email Instructions Text"

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 25]
```diff
@@ -0,0 +1,411 @@
...
+            <TextArea
+                id={'areaOfResearch'}
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:371]: fix

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 26]
```diff
@@ -0,0 +1,411 @@
...
+                value={institution.userInstructions}
+                onChange={(s: string) => this.setState(fp.set(['institution', 'userInstructions'], s))}
+            />
+          </FlexColumn>
+        </FlexRow>
+        {this.state.showBackButtonWarning && <Modal>
+          <ModalTitle>Information not saved</ModalTitle>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:378]: Institution not saved?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 27]
```diff
@@ -0,0 +1,411 @@
...
+          <ModalFooter>
+            <Button onClick={() => this.setState({showBackButtonWarning: false})}
+                    type='secondary' style={{marginRight: '2rem'}}>Finish Saving</Button>
+            <Button onClick={() => this.backNavigate()}
+                    type='primary'>Yes Continue</Button>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:383]: Is this the right text?  I don't have the right link for the mocks so I can't check.

File: ui/src/app/pages/admin/admin-institution-edit.spec.tsx [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+    registerApiClient(InstitutionApi, new InstitutionApiStub());
+  });
+
+  it('should render', async() => {
+    const wrapper = component();
+    await waitOneTickAndUpdate(wrapper);
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('should throw error in case of duplicate display name', async() => {
```
[the remark this answers is not in the document]
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.spec.tsx:32]: Ohh i missed this, its an incomplete test with incorrect description: this is for display Name more than 80 characters

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 29]
```diff
@@ -0,0 +1,411 @@
...
+               inputStyle={{width: '16rem', marginTop: '0.8rem'}}/>}
+            <label style={styles.label}>Agreement Type</label>
+            <Dropdown style={{width: '16rem'}} data-test-id='agreement-dropdown'
+                      placeholder='Your Agreement'
+                      options={this.institutionAgreementTypeOptions}
+                      value={institution.duaTypeEnum}
+                      onChange={v => this.setState(fp.set(['institution', 'duaTypeEnum'], v.value))}/>
+            {institution.duaTypeEnum === DuaType.RESTRICTED && <FlexColumn data-test-id='emailAddress' style={{width: '16rem'}}>
+              <label style={styles.label}>Accepted Email Address</label>
+              <TextArea value={institution.emailAddresses && institution.emailAddresses.join(',\n')}
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:350]: the mocks do not have commas here

File: api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java (hunk 1 above)
Comment (NehaBroad) [reply api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java:158]: Conflict Exception throws 409 exception that i can use in UI to show appropriate message

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 13 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:172]: User Instruction is optional

File: api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java (hunk 1 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java:158]: Good point.  ConflictException is good, let's use that.  But we should include the original exception message in the text.  Maybe something like:
`new ConflictException("DataIntegrityViolationException: " + ex.getMessage())`

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 13 above)
Comment (jmthibault79) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:172]: that is also true for organizationTypeOtherText when type is not Other

Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:172]: After taking another look i am adding userInstruction check as its for fieldNotEdit back button not required field check. I feel i can remove organizationTypeOtherText as it is redundant

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 29 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:350]: Will ask Lou to update, he is aware that  i have added commas to make it easier for us to distinguish between two email addresses/domains

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 27 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:383]: Yes, thats been approved by Lou. Will ask him to update the mock

File: ui/src/app/pages/admin/admin-institution-options.tsx [hunk 30]
```diff
@@ -0,0 +1,12 @@
+import {DuaType, OrganizationType} from 'generated/fetch';
+
+export const DuaTypes =  [{label: 'Master', value: DuaType.MASTER},
+  {label: 'Individual', value: DuaType.RESTRICTED}];
+
+export const OrganizationTypeOptions = [
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-options.tsx:6]: very nice!

File: ui/src/app/pages/admin/admin-institution-options.tsx [hunk 31]
```diff
@@ -0,0 +1,12 @@
...
+  {label: 'Industry', value: OrganizationType.INDUSTRY},
+  {label: 'Academic Research Institution', value: OrganizationType.ACADEMICRESEARCHINSTITUTION},
+  {label: 'Educational Institution', value: OrganizationType.EDUCATIONALINSTITUTION},
+  {label: 'Health Center/ Non-Profit', value: OrganizationType.HEALTHCENTERNONPROFIT},
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-options.tsx:10]: nitpick: consistent spacing around the slash.  Either 'Health Center/Non-Profit' or 'Health Center / Non-Profit'

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 19 above)
Comment (jmthibault79) [reply ui/src/app/pages/admin/admin-institution-edit.tsx]: Better but I think we want to either rename this as "missing" or flip the sign here.

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 32]
```diff
@@ -0,0 +1,422 @@
...
+        showOtherInstitutionTextBox: institutionToEdit.organizationTypeEnum === OrganizationType.OTHER
+      });
+    } else {
+      title = 'Add new Institution';
+      this.setState({institutionMode: InstitutionMode.ADD});
+    }
+
+  }
+
+  get institutionTypeOptions() {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:95]: can remove this now, as well as institutionAgreementTypeOptions

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 33]
```diff
@@ -0,0 +1,422 @@
...
+    }
+    return;
+  }
+
+  // Filter out empty line or empty email addresses like <email1>,,<email2>
+  // Confirm each email domain matches with regex
+  validateEmailDomains() {
+    const invalidEmailDomain = [];
+    const {emailDomains} = this.state.institution;
+    this.state.institution.emailDomains =  emailDomains.filter(emailDomain => emailDomain);
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:148]: this looks like another case of unsafe state setting.  also, what is it intended to do?

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 34]
```diff
@@ -0,0 +1,422 @@
...
+      if (institution.duaTypeEnum === DuaType.MASTER) {
+        institution.emailAddresses = [];
+      } else {
+        institution.emailDomains = [];
+      }
+      if (institution.organizationTypeEnum !== OrganizationType.OTHER) {
+        institution.organizationTypeOtherText = null;
+      }
+    }
+    if (this.props.urlParams.institutionId) {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:230]: better to use the Add vs Edit state here

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 35]
```diff
@@ -0,0 +1,422 @@
...
+      displayName, organizationTypeEnum, duaTypeEnum
+    } = institution;
+    const errors = validate({
+      displayName,
+      'emailAddresses': !this.validateEmailAddressPresence(),
+      'emailDomain': !this.validateEmailDomainPresence(),
+      organizationTypeEnum,
+      duaTypeEnum
+    }, {
+      displayName: {presence: {allowEmpty: false}, length: {maximum: 80, tooLong: 'must be %{count} characters or less'}},
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:294]: this is pretty cool

File: ui/src/app/pages/admin/admin-institution-edit.tsx (hunk 33 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-institution-edit.tsx:148]: remove empty emailDomains "If you want to keep only truthy values you may use .filter(e => e)" however i might use trim as well just to avoid strings like '    '

File: api/src/main/resources/workbench-api.yaml [hunk 36]
```diff
@@ -2347,6 +2347,15 @@ paths:
           description: 'The Institution created by this operation.'
           schema:
             "$ref": "#/definitions/Institution"
+        400:
+          description: Bad request
+          schema:
+            "$ref": "#/definitions/ErrorResponse"
+        409:
+          description: Institution with same display Name already exist
```
Comment (jmthibault79) [inline api/src/main/resources/workbench-api.yaml:2355]: Something like:

DataIntegrityViolationException was thrown, possibly due to an attempt to create a duplicate Institution.

File: ui/src/app/pages/admin/admin-institution-edit.tsx [hunk 37]
```diff
@@ -0,0 +1,410 @@
...
+      if (institution.duaTypeEnum === DuaType.MASTER) {
+        institution.emailAddresses = [];
+      } else {
+        institution.emailDomains = [];
+      }
+      if (institution.organizationTypeEnum !== OrganizationType.OTHER) {
+        institution.organizationTypeOtherText = null;
+      }
+    }
+    if (InstitutionMode.EDIT) {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-institution-edit.tsx:214]: if (institutionEditMode === InstitutionMode.EDIT)
